### PR TITLE
Update vectorDB index to GPU_CAGRA and Milvus image to 2.4.4

### PR DESCRIPTION
**Pull request by nv-nikkulkarni**

### CODE REVIEW

File: deploy/helm/values.yaml [hunk 1]
```diff
@@ -17,7 +17,7 @@ ragServer:
       PROMPT_CONFIG_FILE: "/prompt.yaml"
       APP_VECTORSTORE_URL: "http://milvus:19530"
       APP_VECTORSTORE_NAME: "milvus"
-      APP_VECTORSTORE_INDEXTYPE: CAGRA
+      APP_VECTORSTORE_INDEXTYPE: GPU_SCAGRA
```
Comment (shubhadeepd) [inline deploy/helm/values.yaml]: Typo `CAGRA`

**Review by shubhadeepd — CHANGES_REQUESTED** [review]

**Review by shubhadeepd — APPROVED** [review]

**Comment by droeding** [comment]
# Comprehensive PR Review

**PR**: #9 — feat: Sidecar State Pattern fuer K.O.-Kriterien Persistenz
**Reviewed by**: 5 specialized agents (code-review, error-handling, test-coverage, comment-quality, docs-impact)
**Date**: 2026-04-15

---

## Summary

Clean, well-structured implementation of the Sidecar State Pattern. Core logic (state_extractor.py, models.py) is solid with 21 unit tests. Two HIGH-severity gaps: missing tests for state injection in `context.py` and missing CLAUDE.md documentation updates. No CRITICAL issues.

**Verdict**: `REQUEST_CHANGES`

| Severity | Count |
|----------|-------|
| CRITICAL | 0 |
| HIGH | 2 |
| MEDIUM | 6 |
| LOW | 4 |

*15 raw findings, 12 unique after deduplication*

---

## HIGH Issues

### 1. No test for pipeline_state injection in build_context_for_module()
**Agent**: test-coverage | `src/tender_analysis/context.py:342-348`

The state injection block — which inserts the requirement status table at position 0 of the LLM context — is completely untested. If `sections.insert(0, ...)` were changed to `sections.append(...)`, no test would catch this ordering regression.

<details>
<summary>View recommended test</summary>

```python
def test_build_context_with_pipeline_state():
    """Pipeline state block should be injected at the TOP of context."""
    from tender_analysis.models import PipelineState
    state = PipelineState()
    state.update_requirement("REQ-SICH-001", "ERFUELLT", "technical", "KO")
    result = build_context_for_module(
        context_type=None, config={}, previous_results={},
        company_profile="Bechtle AG Profil", pipeline_state=state,
    )
    assert "SINGLE SOURCE OF TRUTH" in result
    pos_state = result.index("SINGLE SOURCE OF TRUTH")
    pos_profile = result.index("Bechtle AG Profil")
    assert pos_state < pos_profile

def test_build_context_without_pipeline_state():
    """No state block when pipeline_state is None or empty."""
    result = build_context_for_module(
        context_type=None, config={}, previous_results={},
        company_profile="Bechtle AG Profil", pipeline_state=None,
    )
    assert "SINGLE SOURCE OF TRUTH" not in result
```

</details>

---

### 2. tender-analyse/CLAUDE.md missing state management components
**Agent**: docs-impact | `tender-analyse/CLAUDE.md`

New file `state_extractor.py` not listed. `async_pipeline.py` and `context.py` descriptions don't mention state accumulation/injection. "Persistenz & State" section missing `PipelineState`/`RequirementState`.

<details>
<summary>View suggested updates</summary>

- Add `state_extractor.py` to Pipeline-System section
- Update `async_pipeline.py` entry: mention State-Akkumulation
- Update `context.py` entry: mention PipelineState-Injection
- Add `PipelineState`/`RequirementState` to Persistenz & State section

</details>

---

## MEDIUM Issues (Needs Decision)

### 1. `update_requirement` silently discards `req_type` on re-encounter
**Agent**: code-review | `models.py:33-40`

When a requirement is re-assessed with a different type (e.g., PFLICHT → KO), the type is silently ignored. Recommended: monotonic type escalation (KO > PFLICHT > WUNSCH).

**Options**: Fix now (recommended) | Create issue | Skip

<details>
<summary>View fix</summary>

```python
TYPE_PRIORITY = {"KO": 3, "PFLICHT": 2, "WUNSCH": 1, "UNKNOWN": 0}

def update_requirement(self, req_id, status, module, req_type="UNKNOWN"):
    if req_id in self.requirements:
        self.requirements[req_id].status = status
        self.requirements[req_id].handled_in_module = module
        current_priority = TYPE_PRIORITY.get(self.requirements[req_id].type, 0)
        new_priority = TYPE_PRIORITY.get(req_type, 0)
        if new_priority > current_priority:
            self.requirements[req_id].type = req_type
    else:
        self.requirements[req_id] = RequirementState(
            req_id=req_id, type=req_type, status=status, handled_in_module=module
        )
```

</details>

---

### 2. No logging when state extraction finds zero matches
**Agent**: error-handling | `state_extractor.py:25`

Silent return on zero regex matches makes debugging production issues difficult. Recommended: add `logger.debug` with match count.

**Options**: Fix now (recommended) | Skip

---

### 3. Misleading comment about state injection scope
**Agent**: comment-quality | `context.py:342-343`

Comment says "Fuer consolidate und adversarial_review" but all modules receive `pipeline_state`. Recommended: update to match actual behavior.

**Options**: Fix now (recommended) | Skip

---

### 4. No integration test for state accumulation in async_pipeline.py
**Agent**: test-coverage | `async_pipeline.py:447-451`

Guard logic (`if result.success and result.content:`) is untested at integration level. Agent recommends **skip** — function is fully unit-tested, guard is simple, mocking cost is high.

**Options**: Fix now (high effort) | Skip (recommended)

---

### 5. data-models-backend.md — Missing PipelineState/RequirementState
**Agent**: docs-impact | `docs/data-models-backend.md`

New dataclasses not in the data models catalog. Recommended: add 2 table entries.

**Options**: Fix now (recommended) | Create issue

---

### 6. architecture-backend.md — Missing state layer
**Agent**: docs-impact | `docs/architecture-backend.md`

Pipeline/Kontext layer descriptions don't mention state management. Recommended: 2 line edits.

**Options**: Fix now (recommended) | Create issue

---

## LOW Issues

<details>
<summary>View 4 low-priority suggestions</summary>

| # | Issue | Location | Agent | Suggestion |
|---|-------|----------|-------|------------|
| 1 | `to_context_block` renders header-only table when all reqs OFFEN | `models.py:42-59` | code-review | Add early return for no non-OFFEN reqs |
| 2 | Missing generic type annotation on `requirements` dict | `models.py:31` | code-review | Use `dict[str, RequirementState]` |
| 3 | No full-flow test for OFFEN row exclusion | `state_extractor.py:14-16` | test-coverage | Add test with mixed OFFEN/non-OFFEN rows |
| 4 | "(NEU)" tag in pre-existing comment | `context.py:331` | comment-quality | Out of scope — cleanup PR |

</details>

---

## What's Good

- **Tolerant parser design**: Graceful handling of broken LLM output without crashing the pipeline
- **21 tests for 83 lines of code**: Comprehensive coverage with edge cases (broken tables, whitespace, empty outputs)
- **Minimal footprint**: Only 4 source files changed, no unnecessary refactoring
- **Byte-identical prompt additions**: All 6 module prompts have consistent Requirement-Status blocks
- **Smart context injection**: State block at position 0 reinforces Ground Truth semantics
- **Strict regex boundary**: Prevents hallucinated table rows from entering state
- **Pre-existing architecture spec**: Implementation follows documented spec closely

---

## Suggested Follow-up Issues

| Issue Title | Priority | Related Finding |
|-------------|----------|-----------------|
| "Add monotonic type escalation to update_requirement" | P2 | MEDIUM #1 |
| "Add debug logging to state_extractor" | P3 | MEDIUM #2 |
| "Remove outdated (NEU) comment tags" | P3 | LOW #4 |

---

## Next Steps

1. Auto-fix step will address HIGH issues (context.py test + CLAUDE.md docs)
2. Review MEDIUM issues above and decide per item
3. Merge when ready

---

*Reviewed by Archon comprehensive-pr-review workflow*
*Artifacts: `artifacts/runs/9c821abf-7029-496d-9d4b-193f191bc64e/review/`*

**Comment by droeding** [comment]
## 🎯 Workflow Summary

**Plan**: Sidecar State Pattern — K.O.-Kriterien Persistenz
**Workflow ID**: `9c821abf-7029-496d-9d4b-193f191bc64e`
**Status**: ✅ Implementation complete, reviewed, fixes applied

---

### Implementation vs Plan

| Metric | Planned | Actual |
|--------|---------|--------|
| Files created | 2 | 2 |
| Files updated | 10 | 10 |
| Tests added | 21 | 23 (+2 from review fixes) |
| Deviations | - | 1 |

<details>
<summary>📋 Deviation from Plan (1)</summary>

**pipeline_state passed to ALL build_context_for_module calls**
- **Expected**: Only consolidate and adversarial_review should receive `pipeline_state`
- **Actual**: All 4 call sites receive `pipeline_state=pipeline_state`
- **Reason**: Parameter is optional with `None` default. Passing everywhere is simpler — state block only renders when requirements exist. Consolidate runs through the main module loop, so it needs to be there.

</details>

---

### Review Summary

| Severity | Found | Fixed | Remaining |
|----------|-------|-------|-----------|
| CRITICAL | 0 | 0 | 0 |
| HIGH | 2 | 2 | 0 |
| MEDIUM | 6 | 4 | 2 |
| LOW | 4 | 0 | 4 |

**Fixed**: Context injection test (HIGH), CLAUDE.md update (HIGH), debug logging (MED), misleading comment (MED), data-models-backend.md (MED), architecture-backend.md (MED)

---

### 🚀 Quick Wins Before Merge

| Item | Effort | Action |
|------|--------|--------|
| `to_context_block` early return for all-OFFEN | ~3 min | Add `if not filtered: return ""` |
| Type annotation `dict[str, RequirementState]` | ~1 min | Add type hint to `requirements` field |

---

### 📋 Suggested Follow-Up Issues

| Title | Labels |
|-------|--------|
| Add monotonic type escalation to `PipelineState.update_requirement` | `enhancement`, `P2` |
| Remove outdated (NEU) comment tags in context.py | `cleanup`, `P3` |

---

### ⚠️ Deferred Items (NOT Building)

These were **intentionally excluded** from scope:

| Item | Rationale |
|------|-----------|
| Epic 2 (Commercial-Modul) | Stories 2.1–2.5 deferred to separate branch/PR |
| Preis-/Margen-Logik | Nur State-Infrastruktur in diesem PR |
| Prompt 08 (Adversarial Review) | Nur Module 02–07 + 07_consolidate |
| Integration test for async_pipeline state accumulation | Review recommended skip — unit tests sufficient, mocking expensive |

---

**Artifacts**: `/home/demo/.archon/workspaces/droeding/rag/artifacts/runs/9c821abf-7029-496d-9d4b-193f191bc64e/`
